File: app/src/main/java/com/example/buildmypc/ui/parts/parts/Memory.java
package com.example.buildmypc.ui.parts.parts;

import android.os.Parcel;
import android.os.Parcelable;

import org.jetbrains.annotations.Contract;
import org.jetbrains.annotations.NotNull;

public class Memory extends Part implements Parcelable {
	public static final Creator<Memory> CREATOR = new Creator<Memory>() {
		@NotNull
		@Contract("_ -> new")
		@Override
		public Memory createFromParcel(Parcel in) {
			return new Memory(in);
		}

		@NotNull
		@Contract(value = "_ -> new", pure = true)
		@Override
		public Memory[] newArray(int size) {
			return new Memory[size];
		}
	};
	private boolean hasECC;
	private int latencyCAS; // backwards to respect carrot case
	private int ddrGen; // pretty much always four
	private int firstWordLatency; // in nanoseconds
	private String formFactor;
	private boolean heatSpreader;
	private int moduleSize; // in gb
	private int moduleCount; // regular old integer
	private int speed; // in mhz
	private String timing;
	private double voltage;

	public Memory(String model, String manufacturer) {
		super(model, manufacturer);
	}

	public Memory(String model, String manufacturer, boolean hasECC, int latencyCAS, int ddrGen, int firstWordLatency, String formFactor, boolean heatSpreader, int moduleSize, int moduleCount, int speed, String timing, double voltage) {
		this(model, manufacturer);
		this.hasECC = hasECC;
		this.latencyCAS = latencyCAS;
		this.ddrGen = ddrGen;
		this.firstWordLatency = firstWordLatency;
		this.formFactor = formFactor;
		this.heatSpreader = heatSpreader;
		this.moduleSize = moduleSize;
		this.moduleCount = moduleCount;
		this.speed = speed;
		this.timing = timing;
		this.voltage = voltage;
	}

	public Memory(@NotNull Parcel in) {
		super(in.readString(), in.readString());
		hasECC = in.readBoolean();
		latencyCAS = in.readInt();
		ddrGen = in.readInt();
		firstWordLatency = in.readInt();
		formFactor = in.readString();
		heatSpreader = in.readBoolean();
		moduleSize = in.readInt();
		moduleCount = in.readInt();
		speed = in.readInt();
		timing = in.readString();
		voltage = in.readDouble();
	}

	public boolean getHasECC() {
		return hasECC;
	}

	public void setHasECC(boolean hasECC) {
		this.hasECC = hasECC;
	}

	public int getLatencyCAS() {
		return latencyCAS;
	}

	public void setLatencyCAS(int latencyCAS) {
		this.latencyCAS = latencyCAS;
	}

	public int getDdrGen() {
		return ddrGen;
	}

	public void setDdrGen(int ddrGen) {
		this.ddrGen = ddrGen;
	}

	public int getFirstWordLatency() {
		return firstWordLatency;
	}

	public void setFirstWordLatency(int firstWordLatency) {
		this.firstWordLatency = firstWordLatency;
	}

	public String getFormFactor() {
		return formFactor;
	}

	public void setFormFactor(String formFactor) {
		this.formFactor = formFactor;
	}

	public boolean getHeatSpreader() {
		return heatSpreader;
	}

	public void setHeatSpreader(boolean heatSpreader) {
		this.heatSpreader = heatSpreader;
	}

	public int getModuleSize() {
		return moduleSize;
	}

	public void setModuleSize(int moduleSize) {
		this.moduleSize = moduleSize;
	}

	public int getModuleCount() {
		return moduleCount;
	}

	public void setModuleCount(int moduleCount) {
		this.moduleCount = moduleCount;
	}

	public int getSpeed() {
		return speed;
	}

	public void setSpeed(int speed) {
		this.speed = speed;
	}

	public String getTiming() {
		return timing;
	}

	public void setTiming(String timing) {
		this.timing = timing;
	}

	public double getVoltage() {
		return voltage;
	}

	public void setVoltage(double voltage) {
		this.voltage = voltage;
	}

	@Override
	public boolean equals(Object o) {
		if (this == o) return true;
		if (o == null || getClass() != o.getClass()) return false;
		if (!super.equals(o)) return false;

		Memory memory = (Memory) o;

		if (getHasECC() != memory.getHasECC()) return false;
		if (getLatencyCAS() != memory.getLatencyCAS()) return false;
		if (getDdrGen() != memory.getDdrGen()) return false;
		if (getFirstWordLatency() != memory.getFirstWordLatency()) return false;
		if (getHeatSpreader() != memory.getHeatSpreader()) return false;
		if (getModuleSize() != memory.getModuleSize()) return false;
		if (getModuleCount() != memory.getModuleCount()) return false;
		if (getSpeed() != memory.getSpeed()) return false;
		if (Double.compare(memory.getVoltage(), getVoltage()) != 0) return false;
		if (getFormFactor() != null ? !getFormFactor().equals(memory.getFormFactor()) : memory.getFormFactor() != null)
			return false;
		return getTiming() != null ? getTiming().equals(memory.getTiming()) : memory.getTiming() == null;
	}

	@Override
	public int hashCode() {
		int result = super.hashCode();
		long temp;
		result = 31 * result + (getHasECC() ? 1 : 0);
		result = 31 * result + getLatencyCAS();
		result = 31 * result + getDdrGen();
		result = 31 * result + getFirstWordLatency();
		result = 31 * result + (getFormFactor() != null ? getFormFactor().hashCode() : 0);
		result = 31 * result + (getHeatSpreader() ? 1 : 0);
		result = 31 * result + getModuleSize();
		result = 31 * result + getModuleCount();
		result = 31 * result + getSpeed();
		result = 31 * result + (getTiming() != null ? getTiming().hashCode() : 0);
		temp = Double.doubleToLongBits(getVoltage());
		result = 31 * result + (int) (temp ^ (temp >>> 32));
		return result;
	}

	@Override
	public void writeToParcel(@NotNull Parcel dest, int flags) {
		super.writeToParcel(dest, flags);
		dest.writeBoolean(hasECC);
		dest.writeInt(latencyCAS);
		dest.writeInt(ddrGen);
		dest.writeInt(firstWordLatency);
		dest.writeString(formFactor);
		dest.writeBoolean(heatSpreader);
		dest.writeInt(moduleSize);
		dest.writeInt(moduleCount);
		dest.writeInt(speed);
		dest.writeString(timing);
		dest.writeDouble(voltage);
	}
}

File: app/src/main/java/com/example/buildmypc/ui/parts/parts/Memory.kt
package com.example.buildmypc.ui.parts.parts

import android.os.Parcel
import android.os.Parcelable
import android.os.Parcelable.Creator
import org.jetbrains.annotations.Contract
import java.lang.Double.doubleToLongBits

class Memory : Part, Parcelable {
	var hasECC = false
	var latencyCAS // backwards to respect carrot case
			= 0
	var ddrGen // pretty much always four
			= 0
	var firstWordLatency // in nanoseconds
			= 0
	var formFactor: String? = null
	var heatSpreader = false
	var moduleSize // in gb
			= 0
	var moduleCount // regular old integer
			= 0
	private var speed // in mhz
			= 0
	var timing: String? = null
	var voltage = 0.0

	constructor(model: String?, manufacturer: String?) : super(model, manufacturer)
	constructor(
		model: String?,
		manufacturer: String?,
		hasECC: Boolean,
		latencyCAS: Int,
		ddrGen: Int,
		firstWordLatency: Int,
		formFactor: String?,
		heatSpreader: Boolean,
		moduleSize: Int,
		moduleCount: Int,
		speed: Int,
		timing: String?,
		voltage: Double
	) : this(model, manufacturer) {
		this.hasECC = hasECC
		this.latencyCAS = latencyCAS
		this.ddrGen = ddrGen
		this.firstWordLatency = firstWordLatency
		this.formFactor = formFactor
		this.heatSpreader = heatSpreader
		this.moduleSize = moduleSize
		this.moduleCount = moduleCount
		this.speed = speed
		this.timing = timing
		this.voltage = voltage
	}

	constructor(`in`: Parcel) : super(`in`.readString(), `in`.readString()) {
		hasECC = `in`.readBoolean()
		latencyCAS = `in`.readInt()
		ddrGen = `in`.readInt()
		firstWordLatency = `in`.readInt()
		formFactor = `in`.readString()
		heatSpreader = `in`.readBoolean()
		moduleSize = `in`.readInt()
		moduleCount = `in`.readInt()
		speed = `in`.readInt()
		timing = `in`.readString()
		voltage = `in`.readDouble()
	}

	override fun equals(other: Any?): Boolean {
		if (this === other) return true
		if (other == null || javaClass != other.javaClass) return false
		if (!super.equals(other)) return false
		val memory = other as Memory
		if (hasECC != memory.hasECC) return false
		if (latencyCAS != memory.latencyCAS) return false
		if (ddrGen != memory.ddrGen) return false
		if (firstWordLatency != memory.firstWordLatency) return false
		if (heatSpreader != memory.heatSpreader) return false
		if (moduleSize != memory.moduleSize) return false
		if (moduleCount != memory.moduleCount) return false
		if (speed != memory.speed) return false
		if (memory.voltage.compareTo(voltage) != 0) return false
		if (if (formFactor != null) formFactor != memory.formFactor else memory.formFactor != null) return false
		return if (timing != null) timing == memory.timing else memory.timing == null
	}

	override fun hashCode(): Int {
		var result = super.hashCode()
		result = 31 * result + if (hasECC) 1 else 0
		result = 31 * result + latencyCAS
		result = 31 * result + ddrGen
		result = 31 * result + firstWordLatency
		result = 31 * result + if (formFactor != null) formFactor.hashCode() else 0
		result = 31 * result + if (heatSpreader) 1 else 0
		result = 31 * result + moduleSize
		result = 31 * result + moduleCount
		result = 31 * result + speed
		result = 31 * result + if (timing != null) timing.hashCode() else 0
		val temp = doubleToLongBits(voltage)
		result = 31 * result + (temp xor (temp ushr 32)).toInt()
		return result
	}

	override fun writeToParcel(dest: Parcel, flags: Int) {
		super.writeToParcel(dest, flags)
		dest.writeBoolean(hasECC)
		dest.writeInt(latencyCAS)
		dest.writeInt(ddrGen)
		dest.writeInt(firstWordLatency)
		dest.writeString(formFactor)
		dest.writeBoolean(heatSpreader)
		dest.writeInt(moduleSize)
		dest.writeInt(moduleCount)
		dest.writeInt(speed)
		dest.writeString(timing)
		dest.writeDouble(voltage)
	}

	override val paramCount: Int
		get() = 13

	companion object {
		@JvmField
		val CREATOR: Creator<Memory> = object : Creator<Memory> {
			@Contract("_ -> new")
			override fun createFromParcel(`in`: Parcel) = Memory(`in`)

			@Contract(value = "_ -> new", pure = true)
			override fun newArray(size: Int): Array<Memory?> = arrayOfNulls(size)
		}
	}
}